src/invidious/helpers/extractors.cr

```diff
@@ -0,0 +1,400 @@
+      if items_container = renderer_container_contents["shelfRenderer"]?
+        raw_items << renderer_container_contents
+        next
+      elsif items_container = renderer_container_contents["gridRenderer"]?
+      else
+        items_container = renderer_container_contents
+      end
+
+      items_container["items"].as_a.each do |item|
+        raw_items << item
+      end
+    end
+
+    return raw_items
+  end
+end
+
+private struct SearchResultsExtractor < ItemsContainerExtractor
+  def process(initial_data)
+    if target = initial_data["twoColumnSearchResultsRenderer"]?
+      self.extract(target)
+    end
+  end
+
+  private def extract(target)
+    raw_items = [] of Array(JSON::Any)
+    content = target["primaryContents"]
+    renderer = content["sectionListRenderer"]["contents"].as_a.each do |node|
+      if node = node["itemSectionRenderer"]?
+        raw_items << node["contents"].as_a
+      end
+    end
+
+    raw_items = raw_items.flatten
+
+    return raw_items
+  end
+end
+
+private struct ContinuationExtractor < ItemsContainerExtractor
+  def process(initial_data)
+    if target = initial_data["continuationContents"]?
+      self.extract(target)
+    elsif target = initial_data["appendContinuationItemsAction"]?
+      self.extract(target)
+    end
+  end
+
+  private def extract(target)
+    raw_items = [] of JSON::Any
+    if content = target["gridContinuation"]?
+      raw_items = content["items"].as_a
+    elsif content = target["continuationItems"]?
+      raw_items = content.as_a
+    end
+
+    return raw_items
+  end
+end
+
+# Parses an item from Youtube's JSON response into a more usable structure.
+# The end result can either be a SearchVideo, SearchPlaylist or SearchChannel.
+def extract_item(item : JSON::Any, author_fallback : String? = nil,
+                 author_id_fallback : String? = nil)
+  author_fallback = AuthorFallback.new(author_fallback, author_id_fallback)
+
+  # Cycles through all of the item parsers and attempt to parse the raw YT JSON data.
+  # Each parser automatically validates the data given to see if the data is
+  # applicable to itself. If not nil is returned and the next parser is attemped.
+  ITEM_PARSERS.each do |parser|
+    result = parser.process(item, author_fallback)
+    if !result.nil?
+      return result
+    end
+  end
+  # TODO radioRenderer, showRenderer, shelfRenderer, horizontalCardListRenderer, searchPyvRenderer
+end
+
+# Parses multiple items from Youtube's initial JSON response into a more usable structure.
+# The end result is an array of SearchItem.
+def extract_items(initial_data : Hash(String, JSON::Any), author_fallback : String? = nil,
+                  author_id_fallback : String? = nil) : Array(SearchItem)
+  items = [] of SearchItem
+
+  if unpackaged_data = initial_data["contents"]?.try &.as_h
+  elsif unpackaged_data = initial_data["response"]?.try &.as_h
+  elsif unpackaged_data = initial_data["onResponseReceivedActions"]?.try &.as_a.[0].as_h
```

Comment:
According to the documentation the `#as_*?` methods only checks if the underlying type is the requested one and internally it's pretty much just a call to `#is_a?`. 

https://github.com/crystal-lang/crystal/blob/3184e1927ddb88dcaef8ab21c8e3fb3e21944f84/src/json/any.cr#L245-L247

So the try block is needed in this case

Reply:
it calls `#as_h` if the field exists and the underlying type is a `Hash`, otherwise it returns `nil`.

The try block does the same (check if field exists or returns `nil`, and tries to return a `Hash`)

Reply:
> it calls #as_h if the field exists and the underlying type is a Hash, otherwise it returns nil.

No it doesn't. The `#as_*?` methods only checks if the underlying value is the requested type. It doesn't do anything else. 

```crystal
  # Checks that the underlying value is `Array`, and returns its value.
  # Returns `nil` otherwise.
  def as_a? : Array(Any)?
    as_a if @raw.is_a?(Array)
  end
...
  # Checks that the underlying value is `Hash`, and returns its value.
  # Returns `nil` otherwise.
  def as_h? : Hash(String, Any)?
    as_h if @raw.is_a?(Hash)
  end
```
https://github.com/crystal-lang/crystal/blob/3184e1927ddb88dcaef8ab21c8e3fb3e21944f84/src/json/any.cr#L231-L235


In this case, we have to validate that:
1. A specific key exist ( and then retrieve the value)
2. That it is a hash
 
Therefore, we'll have to do an expression in this syntax: 
`initial_data["response"]?.try &.as_h`
If we only did `as_h?` then when the underlying value isn't a "response", it'll cause a KeyError.
`initial_data["response"].as_h?`

Reply:
ok